### PR TITLE
Universal API implementations for Micro-Manager OME-TIFF and NDTiff

### LINKED ISSUE #53: Improve TIFF I/O stack
For now we use a combination of `tifffile`, custom code (multi-page OME-TIFF), and `ndtiff` (pycromanager NDTIFF) to read TIFF-based formats. Given that `tifffile` also supports NDTIFF, there may be room to simplify the reading stack.

Also, we should investigate if `tifffile` and AICSImageIO has improved their OME-TIFF reading performance.

@nclack also mentioned that there are additional strategies we can explore to accelerate TIFF I/O.

### LINKED ISSUE #85: Refactor `_set_mm_meta` in `BaseReader` and its children
This discussion started on #71.

Multiple classes have extremely similar `_set_mm_meta` and other micromanager-related methods that could be simplified and reduce code duplication.

Additionally, why do non-micromanager classes need special micromanager attributes? Shouldn't they live with the rest of the metadata?

### LINKED ISSUE #151: The MicromanagerOmeTiffReader needs methods to retrieve the image plane metadata
The MicromanagerOmeTiffReader needs methods to retrieve the image plane metadata. The internal method `_infer_image_meta` can be modified/extended to retrieve all data in the `MicroManagerMetadata` tif tag as a dictionary given the image PTCZ coordinates.

### LINKED ISSUE #152: The MicromanagerOmeTiffReader needs better definitions or documentation for position attributes
The MicromanagerOmeTiffReader provides `positions`, `stage_positions`, `position_arrays`, and `get_num_positions` attributes / methods. We need better definition and/or documentation for these. Some (like `position_arrays`) should probably be internal and others (like `get_num_positions` which returns `self.positions`) should probably be deprecated.

### LINKED ISSUE #164: Single-page TIFF converter regression?
CI runs after merging #159 has been [failing](https://github.com/czbiohub-sf/iohub/actions/runs/5648283652) for the single-page TIFF converter. However I was not able to reproduce locally. Also that PR shouldn't have changed how single-page TIFFs work (tests passed before merging).

### LINKED ISSUE #180: `iohub info` fails on MM datasets without `xy_pixel_size`
...and `xy_pixel_size` is absent in our 1.3.0 zenodo datasets (and oftentimes from the microscopes if the pixel size isn't calibrated). 

For example:
```
$ iohub info 2022_08_04_recOrder_pytest_20x_04NA/2T_3P_16Z_128Y_256X_Kazansky_1/
Reading file:	 /Users/talon.chandler/Desktop/0.4.0-release/misc/recOrder_test_data/2022_08_04_recOrder_pytest_20x_04NA/2T_3P_16Z_128Y_256X_Kazansky_1
Traceback (most recent call last):
  File "/opt/anaconda3/envs/recorder-dev/bin/iohub", line 8, in <module>
    sys.exit(cli())
  File "/opt/anaconda3/envs/recorder-dev/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/opt/anaconda3/envs/recorder-dev/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/opt/anaconda3/envs/recorder-dev/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/anaconda3/envs/recorder-dev/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/anaconda3/envs/recorder-dev/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/talon.chandler/iohub/iohub/cli/cli.py", line 43, in info
    print_info(file, verbose=verbose)
  File "/Users/talon.chandler/iohub/iohub/reader.py", line 267, in print_info
    reader.xy_pixel_size,
  File "/Users/talon.chandler/iohub/iohub/multipagetiff.py", line 309, in xy_pixel_size
    raise AttributeError("XY pixel size cannot be determined.")
AttributeError: XY pixel size cannot be determined.. Did you mean: '_xy_pixel_size'?
```

### LINKED ISSUE #207: unpin zarr versions
Hi, is it time to revert the change done in #165?

How can I check if `zarr<2.16` is still a requirement?

### PR DESCRIPTION
**Pull request by ziw-liu**
Implements universal API (UAPI) for MMStack (Micro-Manager OME-TIFF) and NDTiff (written by MM and pycro-manager). This PR contains many breaking changes and is geared towards the v0.2.0 release.

### Unresolved issues

- UAPI for OME-NGFF
- #157 (@JoOkuma)

#### API

- UAPI entry point (#40): rename `read_micromanager` to `read_images`
  - Removed `extract_data` and `format` arguments
  - Suggestions to function name are welcome!
  - TODO: include clear control?
- UAPI for MM OME-TIFF
  - Replace `multipagetiff.MicromanagerOmeTiffReader` with `mmstack.MMStack`
    - [x] Closes #152 (@ieivanov)
    - [x] Closes #85 (@JoOkuma)
  - When the channel names from metadata is incomplete (e.g. multi-camera mode on the Dragonfly microscope), the name list will be padded with generic names, e.g. `[..., "Channel2", "Channel3", ...]`. Note that this will be *incorrect* if the partially available channel names do not start with the first channel or if they are not contiguous. A warning will be logged and the user is responsible to correct this downstream.
- UAPI for NDTiff
  - Replace `ndtiff.NDTiffReader` with `ndtiff.NDTiffDataset`
  - When channel names are not available from metadata, fallback to `['Channel0', 'Channel1', ...]`  instead of `['0', '1', ...]`
- OME-TIFF and NDTiff FOV objects now have a `frame_metadata` object as proposed in #155
  - [x] Closes #151 (@ieivanov)
- Non-UAPI readers (MM single-page tiff, v0.1 OME-Zarr, uPTI) are moved into the `_deprecated` namespace
  - Only accessible from specific API (importing corresponding reader classes)
  - Testing is still maintained and there is no immediate plan to remove them
- When ZYX scaling information is not available from OME-TIFF metadata, the reader will fall back to `1.0` and log a warning.
  - [x] Closes #180 (@talonchandler)
  - [x] Replaces #182 (@talonchandler)
- Renamed `convert.TIFFConverter.run()` to `convert.TIFFConverter.__call__()`
 
#### CLI

- Removed support for converting single-page TIFFs
- Removed arguments from the `convert` command
  - `--format` is always auto-inferred
  - `--scale-voxels` is always TRUE
  - `--check-image` is always FALSE
- The chunk size argument `--chunks` now works for both NDTiff and OME-TIFF
  - The default chunk size is ZYX for all input formats
- Avoid using the root logger and set the default logging level to INFO (#188)
  - The logging level can be changed with an environment variable, e.g. `IOHUB_LOG_LEVEL=DEBUG`

#### Documentation

- Updated sphinx theme
- TODO: add version switcher for old releases

### Technical changes

- Use tifffile's `TiffZarrStore` to acheive lazy loading of OME-TIFF. The custom index map parsing is removed.
  - [x] Closes #53
- Use xarray to provide named axes in the TCZYX order. (The `xdata` attribute gives access to the underlying xarray object.)
- Bumps tifffile and zarr-python to latest
  - [x] Closes #207 (@JoOkuma)
  - [x] Closes #164
  - [x] Replaces #208 (@ilan-francisco)
- Reader testing is now done for all the applicable files
  - Replaced session-scoped fixtures in `conftest.py` with globals since parametrization happens before fixture is available

### Notes

Many issues are linked to be closed by this PR.
Reviewers: please check the boxes and leave a sign-off when an issue has been verified to be fixed.

### CODE REVIEW

**Comment by ziw-liu** [comment]
@JoOkuma Not sure if this is going to work for the DaXi data since I don't have 1023x2048x2048 volumes to test with...
Also if you don't like the Dask conversion, the `TiffZarrStore` (in acquisition axes ordering) is available through `MMStack._store`.

**Comment by JoOkuma** [comment]
@ziw-liu, thank you. I'll give it a try soon

**Comment by ziw-liu** [comment]
Fixing docs...

**Comment by ziw-liu** [comment]
@talonchandler channel names are back! Chunk sizes now prints by default for single FOVs and is hidden behind `--verbose` for plates.

**Comment by JoOkuma** [comment]
Hi @ziw-liu , thanks for moving forward with this.

I'll take a look at it soon.

**Comment by ieivanov** [comment]
Very exciting indeed, I'll start testing too.

File: iohub/reader.py
Comment (JoOkuma) [inline iohub/reader.py:48]: Could you check if `len(files)` work?
I think it will return ```TypeError:` object of type 'generator' has no len()```

**Review by JoOkuma — COMMENTED** [review]
Hi @ziw-liu , great PR! I left a few comments.

I notice you're using `array`. Should we adopt this to all other classes too?
I'm unfamiliar with it, but I'm happy to adopt it with our formats if it's zarr-like.

File: iohub/convert.py
Comment (ziw-liu) [reply iohub/convert.py:68]: This currently relies on micro-manager metadata to determine plate layout and write acquisition metadata. But you are right that another FOV mapping can be made to work. We don't currently have any concrete example of that though.

File: iohub/reader.py
Comment (ziw-liu) [reply iohub/reader.py]: Thanks for catching this! Also added a new test.

**Comment by ziw-liu** [comment]
> I notice you're using array. Should we adopt this to all other classes too?

@JoOkuma I think the `ArrayLike` return type is OK. I'm using Xarray since Dask has to be used for TIFFs anyway. In other cases Dask may not be ideal for performance reasons.

**Comment by ieivanov** [comment]
I really like this implementation! I'll post messages as I continue reading and testing and I'll provide an official review at the end.

> Suggestions to function name are welcome!

Renaming suggestion - currently we have the following class names

`BaseFOV` : `BaseFOVMapping`
`MicroManagerFOV` : `MicroManagerFOVMapping`
`MMOmeTiffFOV` : `MMStack`
`NDTiffFOV` : `NDTiffDataset`

it may be more "symmetric" to name the classes as follows:

`BaseFOV` : `BaseFOVStack`
`MicroManagerFOV` : `MicroManagerFOVStack`
`MMOmeTiffFOV` : `MMStack`
`NDTiffFOV` : `NDTiffStack`

`stack` could be replaced with `dataset`, if you prefer. One more iteration could be:

`BaseFOV` : `BaseFOVStack`
`MicroManagerFOV` : `MicroManagerFOVStack`
`OMETiffFOV` : `OMETiffStack`
`NDTiffFOV` : `NDTiffStack`

`OMETiffFOV` inherits from `BaseFOV` and `OMETiffStack` inherits from `MicroManagerFOVStack`. We likely won't be reading other types of ome.tif files.

**Comment by ziw-liu** [comment]
@ieivanov with `MM` in the name of OME-TIFF reader classes I'm hoping to note that there is no support for a spec-adherent OME-TIFF dataset that is not the direct output of MM.

**Review by ieivanov — CHANGES_REQUESTED** [review]
Again, I think this is a great improvement to `iohub`.

Before merging, we should update `readme.md` to reflect the new `read_images` API. It would also be very helpful to include examples of the new slicing and string-indexing that xarray enables. Most/all of the official `iohub` examples currently focus on the zarr reader/writer.

It would also be very helpful to include `__repr__` methods for `MicroManagerFOV`, `MicroManagerFOVMapping`, `NDTiffFOV`, and `NDTiffDataset`, largely providing the same information as the respective `xarray` `__repr__` methods. `_repr_html_` methods would be a cherry on top.

File: iohub/ndtiff.py
Comment (ieivanov) [inline iohub/ndtiff.py:194]: Should this be a private property? 

NDTiff allows for arbitrary axis names. Does this mean that we are not supporting axis names outside of the ones defined here? That would be OK with me.

Comment (ziw-liu) [reply iohub/ndtiff.py:194]: I didn't know that these can be arbitrary strings. Then how can we support that if e.g. the channel axis is not named 'channel'?

**Comment by ziw-liu** [comment]
> It would also be very helpful to include examples of the new slicing and string-indexing that xarray enables. Most/all of the official iohub examples currently focus on the zarr reader/writer.

All of the example scripts are about NGFF or CC. That's because these are the formats we can write with iohub so test data doesn't need to be downloaded during documentation building. We can ask developers to always run pytest before building docs so the test data can be reused. But this will likely be a fragile hack.

@JoOkuma any ideas?

**Comment by ieivanov** [comment]
Access to the MM summary metadata should be exposed for `NDTiffDataset` and `MMStack`. Currently, the summary metadata is available through the `mm_meta` property (which holds all MM metadata). I think `MicroManagerFOVMapping
` should include a property `summary_metadata` which returns `mm_meta['Summary']`. `MicroManagerFOVMapping.mm_meta` should probably also be renamed to `micromanager_metadata`.

File: iohub/ndtiff.py
Comment (ieivanov) [reply iohub/ndtiff.py:196]: Yeah, NDTiff would allow you to create for example `("position", "time", "lf_channels", "fl_channels", "z", "y", "x")`. I think it's OK if we don't support that. I was mostly arguing for `ndtiff_axes` to be a private property, i.e. `_ndtiff_axes`. We could check that the ndtiff dataset axes are in `_ndtiff_axes` and throw a "not supported" error if that's not the case.

Comment (ziw-liu) [reply iohub/ndtiff.py]: https://github.com/czbiohub-sf/iohub/pull/185/commits/a6e13a8ae2a76d8e6e02afec4ad91d0debbe59f8

**Comment by ziw-liu** [comment]
@ieivanov I opened #212. Let's handle that in a new PR.

**Comment by mattersoflight** [comment]
@ziw-liu fantastic work across all the layers of `iohub`. It looks like some of the open issues (#71 , #207 , #208 ) are quick to handle. Please test/fix the code if the original authors are busy and close them. 

I suggest merging when these features are implemented and tested with real data:
* iohub convert can chunk by z (looks like that was done in another PR).
* Unified API is documented on README so that power users/reviewers begin to switch and can provide feedback aligned with future development.

I think the following can be future/smaller PRs that immediately follow this:
* Proper metadata parsing across the formats we use requires a lot more thought. How we handle metadata depends on how we will consume metadata - in the viewer, through API, etc.  One PR per format would make sense to iterate through the metadata handling systematically. 
* User friendly examples.

I don't have input on when it is best to rename methods.  Since many users are now finding this PR valuable and will provide input as they use the code, it is smart not to delay the merging.

**Comment by ziw-liu** [comment]
@ieivanov Is there any blocking issues left before we are ready to merge this?

**Comment by ieivanov** [comment]
One question / request before I approve - when are the [docs](https://czbiohub-sf.github.io/iohub/main/api.html) rebuilt? It looks like only on a push to `main`? It will be very helpful to have docs built for this PR (or the `ometiff-uapi` branch) so that users can have easy access to the updated API.

**Comment by ziw-liu** [comment]
> One question / request before I approve - when are the [docs](https://czbiohub-sf.github.io/iohub/main/api.html) rebuilt? It looks like only on a push to `main`? It will be very helpful to have docs built for this PR (or the `ometiff-uapi` branch) so that users can have easy access to the updated API.

Currently it is only built when pushing to main. I can attempt to also build for this branch, but we won't know if it is going to work (before we actually merge this).

**Comment by ziw-liu** [comment]
@ieivanov I propose that we handle the docs preview in a separate PR so we can easily revert the change after the transition.

**Review by ieivanov — APPROVED** [review]
Let's get this PR merged. A few outstanding non-blocking issues from my point of view based on our discussion so far:

- [ ] I suggest renaming classes as in https://github.com/czbiohub-sf/iohub/pull/185#issuecomment-1955724023. This does not introduce any use-facing changes, but I think would improve the readability of the code.
- [ ] build docs preview showing the new UAPI, to be resolved in a different PR

**Comment by JoOkuma** [comment]
> `stack` could be replaced with `dataset`, if you prefer. One more iteration could be:

I like @ieivanov's suggestion; I think `dataset` would be more appropriate because `stack` is a "well-defined" data structure, which is not what we use for these classes.

**Review by JoOkuma — COMMENTED** [review]
Once again, I would like to thank @ziw-liu for this PR. It's a huge change and will improve our implementation of downstream image processing/analysis routines.

It closes the Zarr and file issues we raised before.

My concerns are:
 
- The primary usage of `xdata` on the new FOV classes, which deviates from the common API. From what I chatted with @ziw-liu, this is a requirement for the TIFF data (so it has array-like behavior? I don't remember quite well).
    I'm worried some routines start depending on it and, therefore, are not universal, which is the main goal of this PR.
   This method could be added to `BaseFOV` so every class must implement its own `xdata` method; the downside is `BaseFOV` starts deviating slightly from being an `ArrayLike` object.
- The convert function being TIFF specific. The original vision of the universal API was that the conversion function would be `BaseFOVMapping -> BaseFOVMapping` --- despite, in practice, the output type being a subset of all supported types because some are read-only.
    This leads me to think that the API isn't universal yet.
    I understand that MicroManager-specific features are not available in other classes, but I think these could be wrapped into existing `BaseFOVMapping` methods; if that's not the case, the `BaseFOVMapping` could be improved.

File: iohub/convert.py
Comment (ziw-liu) [reply iohub/convert.py:21]: Thanks for the catch! This is actually a bug. Fixed in https://github.com/czbiohub-sf/iohub/pull/185/commits/e8fe720519788cb8585a71d5fb54bbe7c2ffe3ad. Now this file is 95% covered.

**Comment by ziw-liu** [comment]
@JoOkuma Can #85 be closed by this?

**Comment by ziw-liu** [comment]
>The primary usage of xdata on the new FOV classes, which deviates from the common API. From what I chatted with @ziw-liu, this is a requirement for the TIFF data (so it has array-like behavior? I don't remember quite well).
I'm worried some routines start depending on it and, therefore, are not universal, which is the main goal of this PR.
This method could be added to BaseFOV so every class must implement its own xdata method; the downside is BaseFOV starts deviating slightly from being an ArrayLike object.

`xarray.DataArray` is also `ArrayLike`. In the case of TIFF readers it is only to label the axes of the underlying `dask.Array`. I think the spirit of UAPI was that one can call `napari.add_image` with any `BaseFOV` child.

**Comment by JoOkuma** [comment]
> @JoOkuma Can #85 be closed by this?

@ziw-liu , I think so.

**Comment by JoOkuma** [comment]
> > The primary usage of xdata on the new FOV classes, which deviates from the common API. From what I chatted with @ziw-liu, this is a requirement for the TIFF data (so it has array-like behavior? I don't remember quite well).
> > I'm worried some routines start depending on it and, therefore, are not universal, which is the main goal of this PR.
> > This method could be added to BaseFOV so every class must implement its own xdata method; the downside is BaseFOV starts deviating slightly from being an ArrayLike object.
> 
> `xarray.DataArray` is also `ArrayLike`. In the case of TIFF readers it is only to label the axes of the underlying `dask.Array`. I think the spirit of UAPI was that one can call `napari.view_image` with any `BaseFOV` child. Edit: IIRC we decided that the full array interface is too broad to implement.

Ok, now that I applied closer attention I understand that you're mainly using `xdata` inside the `__getitem__` methods.
I was a bit concerned because of 

```python
to_zarr(fov.xdata.data.rechunk(self.chunks), zarr_img)
```

But there isn't an easy way around it.

@ziw-liu , I'm very happy with the PR, and for me it's ready to merge and remaining tasks could be converted to issues.

**Comment by ziw-liu** [comment]
@ieivanov @JoOkuma @talonchandler Thanks for the reviews!

**Pull request merged**
